c10/cuda/CUDAStream.cpp

```diff
@@ -304,7 +306,26 @@ LeakyStreamInternals* CUDAStream_internals(CUDAStream s) {
   }
 }
 
+void check_assert_state(const LeakyStreamInternals* ptr) {
+  if (!ptr->assert_state) {
+    return;
+  }
+
+  if (ptr->assert_state->error != 0) {
+    CUDAGuard device_guard(ptr->device_index);
+
+    // wait for kernel to complete
+    C10_CUDA_CHECK(cudaDeviceSynchronize());    
```

Comment:
Naw man, you can't do this! Now you're synchronizing after every kernel again. The point is that checking for failures has to be opportunistic in the same way that `cudaGetLastError` is: if the error flag happens to be set, now you can do a synchronize and grab all the information you need--if it's not, you gotta wait until next time to try.

(One complication is making sure, if we manage to queue more kernels before we discover the previous one is dead, we do something reasonable in that case...)

Reply:
The `cudaDeviceSynchronize()` call is only made when `error != 0`, e.g. after a device side assert has failed and after the write operation to the pinned memory becomes visible to the host. The device sync is done to wait for the (failed) kernels to complete in order to get all information and throw an exception. E.g. otherwise the effect of the `atomicCAS` operation might become 'visible' before the other data has been transfered, e.g. the device-side string-copy operations for the. It would be possible to do everything completely without device-sync but that would get a bit more complicated (e.g. similar to writing a custom host/device spanning critical-section).

Reply:
(It should definitely be possible to launch kernels in parallel as it was before, the sync is only done in those cases that before ended up in the evil "Device-side assert triggered" state...)

Reply:
Ah, you're right!